### PR TITLE
reduce lock contention on latest_root

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
latest_root is called with a lock on the index for every account load. This code could be tighter. We could get the read lock once and keep it if we are able to reduce the amount of time we need to hold it open. We could also check ancestors, which is a smaller set of slots first, before checking the more expensive, 400k roots hashset.

#### Summary of Changes
rework the loop.
Fixes #

### CODE REVIEW

File: runtime/src/accounts_index.rs [hunk 1]
```diff
@@ -750,30 +750,35 @@ impl<T: 'static + Clone + IsCached + ZeroLamport> AccountsIndex<T> {
...
         let mut current_max = 0;
         let mut rv = None;
-        for (i, (slot, _t)) in slice.iter().rev().enumerate() {
-            if *slot >= current_max && self.is_ancestor_or_root(*slot, ancestors, max_root) {
-                rv = Some((slice.len() - 1) - i);
-                current_max = *slot;
+        if let Some(ancestors) = ancestors {
+            if !ancestors.is_empty() {
+                for (i, (slot, _t)) in slice.iter().rev().enumerate() {
+                    if *slot >= current_max && ancestors.contains_key(slot) {
```
Comment (jeffwashington) [inline runtime/src/accounts_index.rs:756]: @sakridge isn't *slot > current_max ok here? What conditions would cause = to be useful? If we already found a root or ancestor at a slot = to the current slot we're looking for, isn't that good enough? Do we expect multiple items in the slice to have the same slot #?

Comment (sakridge) [reply runtime/src/accounts_index.rs:756]: `slot > current_max` should be ok as well. I don't believe there should be duplicates in the list.

**Comment by sakridge** [comment]
Believe it does the right thing. It would be nice to see how fast it is and in what scenarios, like if we have a large ancestors list vs. small and/or a large slot_list set vs. 1 or 2. Also, how contended the roots_tracker lock is. Because I think we just write to that once per slot to update the root, although we do probably read it a lot from banking/clean/rpc

**Comment by jeffwashington** [comment]
I measure read().unwrap().contains to be 5x slower than already_unlocked.contains().

[Replacing HashSet with BitVec](https://github.com/solana-labs/solana/pull/16310) AND getting re-using the read lock measures 10x faster.

**Comment by ryoqun** [comment]
@jeffwashington (cc: @sakridge ) oh, you're tackling lock contention? I don't have much context here, but I generally tried to fix all the slow lock contentions by superior third-party lock library than `std` ones: https://github.com/solana-labs/solana/pull/14578 `std` ones are kind of slow. That pr got pushed out from my work queue sadly, happy to help if you're interested in taking over. :) I wonder  we could gain a cheap win from just the library switch.

**Comment by jeffwashington** [comment]
One line of thinking is that when we use a BitVec instead of HashSet, then .contains() will be small constant time. Ancestors is lock free. So, any call that results in us finding the slot in the ancestors list will mean we never have to get the read lock to the roots. If we have a very long slot list, we will check the ancestors list for each slot. But, with that check being lock free and small constant time, this should be negligible. Then, when we lock the roots, we keep it locked and we again have small constant time checks for each item in the slot list. With BitVec, the # of items in the Ancestors list or Roots list is irrelevant. So, worst case due to this change is long slot list, > 0 ancestors, but no matches in ancestors. We effectively waste time checking ancestors for each slot. Since this is lock free, low constant time per check, this seems negligible.

**Comment by jeffwashington** [comment]
> @jeffwashington (cc: @sakridge ) oh, you're tackling lock contention? I don't have much context here, but I generally tried to fix all the slow lock contentions by superior third-party lock library than `std` ones: #14578 `std` ones are kind of slow. That pr got pushed out from my work queue sadly, happy to help if you're interested in taking over. :) I wonder we could gain a cheap win from just the library switch.

I'm finding things that make poh, hashing, or replay slower than it needs to be ;-) I ran into these 2 code paths. I also added a read only accounts cache last week. With that cache populated, account load time became dominated by this code checking for is_root.

**Comment by jeffwashington** [comment]
oh, and the longer this code path takes, the longer we hold the index locks open, as far as I can tell. And, I believe we do have index lock contention.

**Comment by jeffwashington** [comment]
> @jeffwashington (cc: @sakridge ) oh, you're tackling lock contention? I don't have much context here, but I generally tried to fix all the slow lock contentions by superior third-party lock library than `std` ones: #14578 `std` ones are kind of slow. That pr got pushed out from my work queue sadly, happy to help if you're interested in taking over. :) I wonder we could gain a cheap win from just the library switch.

I tagged myself on that pr. I will be glad to add that to my queue.

**Comment by ryoqun** [comment]
> I ran into these 2 code paths. I also added a read only accounts cache last week. With that cache populated, account load time became dominated by this code checking for is_root.

sounds cool with pretty finished progress. so, you aren't seeing large portion of memcopies are gone now?: https://github.com/solana-labs/solana/issues/14523#issuecomment-765800757

File: runtime/src/accounts_index.rs (hunk 1 above)
Comment (jeffwashington) [reply runtime/src/accounts_index.rs:756]: I worried about this... A problem is a test that uses slot 0. I thought this might be a corner case with unsigned and tests. -1 is such a convenient number but shockingly missing from unsigned types. ;-)

Comment (sakridge) [reply runtime/src/accounts_index.rs:756]: I see, yea it should find slot=0 if it exists.

File: runtime/src/accounts_index.rs [hunk 2]
```diff
@@ -750,30 +750,35 @@ impl<T: 'static + Clone + IsCached + ZeroLamport> AccountsIndex<T> {
...
+                        rv = Some(i);
+                        current_max = *slot;
+                    }
+                }
             }
         }
 
-        rv
-    }
+        let max_root = max_root.unwrap_or(Slot::MAX);
```
Comment (sakridge) [inline runtime/src/accounts_index.rs:764]: I think we could return early here if `rv.is_some()`

**Review by sakridge — APPROVED** [review]

File: runtime/src/accounts_index.rs (hunk 2 above)
Comment (jeffwashington) [reply runtime/src/accounts_index.rs:764]: https://github.com/solana-labs/solana/pull/16378
